src/node_modules.cc

```diff
@@ -0,0 +1,442 @@
+
+  std::string input;
+  if (ReadFileSync(&input, path.data()) < 0) {
+    return std::nullopt;
+  }
+
+  std::string_view json_string(input);
+  if (input.size() >= 3 && 0 == memcmp(input.data(), "\xEF\xBB\xBF", 3)) {
+    json_string.remove_prefix(3);  // Skip UTF-8 BOM.
+  }
+
+  simdjson::ondemand::document document;
+  simdjson::ondemand::object main_object;
+  simdjson::error_code error =
+      binding_data->json_parser.iterate(json_string, json_string.max_size())
+          .get(document);
+
+  const auto throw_invalid_package_config = [error_context, path, realm]() {
+    if (error_context == nullptr) {
+      THROW_ERR_INVALID_PACKAGE_CONFIG(
+          realm->isolate(), "Invalid package config %s.", path.data());
+    } else if (error_context->base.has_value()) {
+      auto file_url = ada::parse(error_context->base.value());
+      CHECK(file_url);
+      auto file_path = url::FileURLToPath(realm->env(), *file_url);
+      CHECK(file_path.has_value());
+      THROW_ERR_INVALID_PACKAGE_CONFIG(
+          realm->isolate(),
+          "Invalid package config %s while importing \"%s\" from %s.",
+          path.data(),
+          error_context->specifier.c_str(),
+          file_path->c_str());
+    } else {
+      THROW_ERR_INVALID_PACKAGE_CONFIG(
+          realm->isolate(), "Invalid package config %s.", path.data());
+    }
+
+    return std::nullopt;
+  };
+
+  if (error || document.get_object().get(main_object)) {
+    return throw_invalid_package_config();
+  }
+
+  simdjson::ondemand::raw_json_string key;
+  simdjson::ondemand::value value;
+  PackageConfig package_config{};
+
+  for (auto field : main_object) {
+    // Throw error if getting key or value fails.
+    if (field.key().get(key) || field.value().get(value)) {
+      return throw_invalid_package_config();
+    }
+
+    if (key == "name") {
+      std::string_view name;
+      if (!value.get_string().get(name)) {
+        package_config.name = std::string(name);
+      }
+    } else if (key == "main") {
+      std::string_view main;
+      if (!value.get_string().get(main)) {
+        package_config.main = std::string(main);
+      }
+    } else if (key == "exports") {
+      simdjson::ondemand::json_type exports_type;
+      if (!value.type().get(exports_type)) {
+        std::string_view exports;
+        switch (exports_type) {
+          case simdjson::ondemand::json_type::object: {
+            simdjson::ondemand::object exports_object;
+            if (!value.get_object().get(exports_object) &&
+                !exports_object.raw_json().get(exports)) {
+              package_config.exports = std::string(exports);
+            }
+            break;
+          }
+          case simdjson::ondemand::json_type::array: {
+            simdjson::ondemand::array exports_array;
+            if (!value.get_array().get(exports_array) &&
+                !exports_array.raw_json().get(exports)) {
+              package_config.exports = std::string(exports);
+            }
```

Comment:
There are many ways to solve these problems, but I'd prefer something like...

```C++
if (value.get_array().get(exports_array) ||
                exports_array.raw_json().get(experts)) {
  // error
}
package_config.exports = std::string(exports);
```

See other proposals in other comments.